tracing/tracing.go

```diff
@@ -37,14 +37,27 @@ func InstrumentHTTP(h http.Handler) http.Handler {
 		}
 		ctx = slog.NewContext(ctx, log)
 
-		log.Debug("handling request", "url", req.URL, "method", req.Method)
+		httpRequest := []any{
+			"method", req.Method,
+			"url", req.URL.String(),
+			"request_size", req.ContentLength,
+			"user_agent", req.UserAgent(),
+			"protocol", req.Proto,
+		}
+
+		log.Debug("handling request", slog.Group("httpRequest", httpRequest...))
 		resWriter := &responseWriter{
 			ResponseWriter: res,
 		}
 		start := time.Now()
 		defer func() {
 			elapsed := time.Since(start)
-			log.Info("handled request", "url", req.URL, "method", req.Method, "status", resWriter.status, "elapsed", elapsed)
+			httpRequest = append(httpRequest,
+				"status", resWriter.status,
+				"response_size", resWriter.contentLength,
+				"elapsed", elapsed.String(),
+			)
+			log.Info("handled request", slog.Group("httpRequest", httpRequest...))
```

Comment:
Unfortunately, I couldn't find a way to rename the group `httpRequest`. I want to have it here `http_request` and replace only for gcloud

Reply:
This is odd, you can't intercept the "http_request" and then rename it to "httpRequest" ? (this is usually possible for any field) I never used Group...maybe define a struct with all the fields you want to log and then calling it here like:

```go
httpReq := httpRequest{
   Status: etc,
   Lala: etc,
}
log.Info("handle request", "http_request", httpReq)
```

Just wondering what could be changing the behavior and not allowing you to change the key.

Reply:
Nope: the signature is for `ReplaceAttr` is `func(groups []string, a slog.Attr) slog.Attr`, I receive the list of groups, but I cannot return it... 

Also from https://pkg.go.dev/log/slog@go1.21.1#HandlerOptions

> 	// The first argument is a list of currently open groups that contain the
	// Attr. It must not be retained or modified. ReplaceAttr is never called
	// for Group attributes, only their contents.

Yep, I can check if this would work, because then is not a group but a single field!

Reply:
If you want to keep things dynamic, I think this also should work (but never used it):

```go
httpReq := map[string]any{
    "a": "b"
}
log.Info("handle request", "http_request", httpReq)
```

Reply:
Not sure, I would need to test. But it seems to me, that it'll be `{"http_request": "{\"a\":\"b\"}"`, like a json inside of the string, not as different fields.

Reply:
I had a use-case on search that I wanted to know what was the log level set. Maybe an approach is being able to get those information. Which is the log level/format, then I could dynamically set the right field. Just an idea

Reply:
It does work with the map when the output is JSON: https://go.dev/play/p/bKyMG2jIFc3
It gets uglier with the text handler (prints a map on the output, instead of the nice grouping).
It is just a little odd how grouping seems redundant with having a struct/map..at least for simple scenarios like these (there is the whole WithGroup that then allows some more complex idioms).

Reply:
Playing around further with this: https://go.dev/play/p/tSORDXDZJ1Q
seems like a rewriting using the Attribute key is possible, instead of grouping (but maybe there is some benefit to grouping that Im missing..Im not very used to this concept, I think slog is the first time I see this Group idea.. usually you only have key/values).

Reply:
> Nope: the signature is for `ReplaceAttr` is `func(groups []string, a slog.Attr) slog.Attr`, I receive the list of groups, but I cannot return it...
> 
> Also from https://pkg.go.dev/log/slog@go1.21.1#HandlerOptions
> 
> > // The first argument is a list of currently open groups that contain the
> > // Attr. It must not be retained or modified. ReplaceAttr is never called
> > // for Group attributes, only their contents.
> 
> Yep, I can check if this would work, because then is not a group but a single field!

Yeah groups can't be changed...this looks like a reason to not using grouping since we deliberately want to change the main object key name (the root httpRequest).

Reply:
> Thinking better, that's what I'd expect in the struct case as well.

yeah I tested with maps because I never used maps and was curious, it definitely works well with structs (for the JSON handler at least).

Reply:
> Not sure, I would need to test. But it seems to me, that it'll be `{"http_request": "{\"a\":\"b\"}"`, like a json inside of the string, not as different fields.

No it won't, I already use that a lot on our services with structs + I just tested with maps. It handles structured data well, it won't marshal it as a single huge string.

Here is an example with structs: https://go.dev/play/p/r2KCl6H4wPy

Reply:
Not behaving this way would make structured logging remarkably hard, makes perfect sense to handle well maps/structs and leverage their structures on the log entry.

Reply:
> I had a use-case on search that I wanted to know what was the log level set. Maybe an approach is being able to get those information. Which is the log level/format, then I could dynamically set the right field. Just an idea

I don't get the problem/solution here (why do you need to know the level and why changes on level would change fields that are set :thinking: ...I get it for the format, but that is handled by having a customized handler etc).

Reply:
Yep, than seems to be a way. I'll try to do those changes and see how it goes

Reply:
It only prints everything as a single string on the TextHandler. If you want a nice output when using the TextHandler then it seems grouping is the only way to go (matches what I saw on the docs too). But I would not optimize for that, the output with the TextHandler is nice enough, and the JSON one works perfectly (for indexing/searching since it preserves the structure/fields).

https://go.dev/play/p/243w0A47rfr